design-documents/service-isolation/authentication-authorization.md

```diff
@@ -0,0 +1,80 @@
+
+There are few multiple strategies how to build authentication and authorization for services:
+    
+1. BFF/Monolith performs authentication and authorization 
+     - Pros: a single entry point for permissions validation, no code duplication
+     - Cons: services should trust BFF/monolith
+2. A service performs authentication and authorization
+     - Pros: services control all the flow of roles and privileges
+     - Cons: the same logic is duplicated in multiple services
+3. BFF/Monolith performs authentication, a service performs authorization
+    - Pros: single entry point for authentication, services control roles & privileges
+    - Cons: shared roles & privileges data between services
+4. BFF/Monolith performs authentication and authorization, a service uses a "secure" token
+    - Pros: a single entry point for credentials and permissions validation, more secure communication, avoids
+[Confused Deputy Problem](https://en.wikipedia.org/wiki/Confused_deputy_problem)
+    - Cons: additional logic to validate a "secure" token (could be extracted to a separate service)
+
+According to described above strategies, the last approach looks like more convenient. It allows building authentication
+& authorization flow with a single source of roles & privileges and provides additional secure communication for separate
+services.
+
+### Token-based Auth
+
+Token-based auth allows users to use credentials like username and password to obtain a token which allows communicating
+with the system without using credentials.
+As Monolith/BFF performs authentication & authorization token-based auth allows using tokens for a secure communication
+between a user, Monolith/BFF, and services.
+
+[JWT](https://jwt.io/) could be used as a "secure" token implementation and the main benefits are:
+* Allows using adaptive structure to pass permissions between services
+* Could be encrypted (JWE) or just signed (JWS)
+* As it's RFC it has implementations for most of the modern programming languages, so there is no a hard
+dependency on some stack for the service implementation
+    
+## General Schema
+
+As Magento should provide [multiple BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends)
+it makes sense to extract authentication and authorization flow to a separate `Auth Service` to avoid credentials
+validation and duplication of roles & privileges data. And the general schema might look like this:
+
+![Authentication and Authorization General Schema](auth-general-schema.png)
+
+And the basic scenario might be as following:
+
+![Auth Basic Scenario](auth-basic-scenario.png)
+
+Also, there are a few variants of the proposed schema like to have separate services for token validation and renewal:
+
+![Token Validation Service](token-validation-service.png)
+
+Now, each service validates a token through `Token Service` it reduces logic duplication but adds additional requests
```

Comment:
> it reduces logic duplication

Logic duplication can be reduced by creating a shared library that will be installed on all services. This will reduce number of network calls, and also will allow to rollout updates of the library gradually, if necessary (though this shouldn't be a frequent task).
Having a single service, it becomes a bottleneck, so if it goes down (or is in upgrade process), all services will be blocked.

Reply:
As at least now, all our services will be on the same stack - a separate validation service doesn't make sense. But potentially, services might be implemented using a different stack, that's why I've described the additional approach. For now shared library it's ok.